spinalcordtoolbox/scripts/sct_dmri_display_bvecs.py

```diff
@@ -52,37 +61,82 @@ def get_parser():
+    # plt.axis('equal')
     plt.xlabel(xlabel)
     plt.ylabel(ylabel)
-    plt.xlim([-1, 1])
-    plt.ylim([-1, 1])
+    plt.xlim([-max(bvals), max(bvals)])
+    plt.ylim([-max(bvals), max(bvals)])
     plt.grid()
 
+
+def create_custom_legend(fig, shell_colors, bvals):
+    """
+    Create single custom legend for whole figure
+    :param fig: figure the legend will be created for
+    :param shell_colors: dict with b-values and colors
+    :param bvals: ndarray with all b-values
+    :return:
+    """
+
+    # count number of bvals for individual shells
+    bvals_per_shell = pd.Series(bvals[bvals > bzero]).value_counts()
+
+    # Create single custom legend for whole figure with several subplots
+    # Loop across legend elements
+    lines = [Line2D([0], [0], color=color, marker='o', markersize=8, alpha=0.7, linestyle='')
+             for color in shell_colors.values()]
+    labels = [f'b-values = {shell} (n = {bvals_per_shell[shell]})'
+              for shell in shell_colors.keys()]
+
+    # Insert legend below subplots, NB - this line has to be run after the plt.tight_layout() which is called before
+    # function call
+    legend = fig.legend(lines, labels, loc='lower left', bbox_to_anchor=(0.1, 0),
+                        bbox_transform=plt.gcf().transFigure, ncol=len(lines), fontsize=10)
+    # Change box's frame color to black
+    frame = legend.get_frame()
+    frame.set_edgecolor('black')
+    # tight layout of whole figure and shift master title up
+    fig.tight_layout()
+    fig.subplots_adjust(top=0.88, bottom=0.1)
+
+
 # MAIN
 # ==========================================================================================
 
-
 def main(argv=None):
     parser = get_parser()
     arguments = parser.parse_args(argv)
     verbose = arguments.v
     set_global_loglevel(verbose=verbose)
 
     fname_bvecs = arguments.bvec
-
-    # Read bvecs
-    bvecs = read_bvals_bvecs(fname_bvecs, None)
-    bvecs = bvecs[0]
+    fname_bvals = arguments.bval
+    # Read bvals and bvecs files (if arguments.bval is not passed, bvals will be None)
+    bvals, bvecs = read_bvals_bvecs(fname_bvals, fname_bvecs)
     # if first dimension is not equal to 3 (x,y,z), transpose bvecs file
-    if not bvecs.shape[0] == 3:
+    if bvecs.shape[0] != 3:
         bvecs = bvecs.transpose()
-    x, y, z = bvecs[0], bvecs[1], bvecs[2]
+    # if bvals file was not passed, create dummy unit bvals array (necessary fot scatter plots)
+    if bvals is None:
+        bvals = np.repeat(1, bvecs.shape[1])
+    # multiply unit b-vectors by b-values
+    x, y, z = bvecs[0] * bvals, bvecs[1] * bvals, bvecs[2] * bvals
+
+    # Set different color for each shell (bval)
+    shell_colors = {}
+    # Create iterator with different colors from brg palette
+    colors = iter(cm.brg(np.linspace(0, 1, len(np.unique(bvals)))))
+    for unique_bval in np.unique(bvals):
+        # skip b=0
+        if unique_bval < bzero:
```

Comment:
This suggestion is outside the scope of this PR, but it's a tiny change that might be nice to do while we're here.

Since `bzero` is a module-scoped (global) constant, what do you think of renaming it to `BZERO_THRESH`? While reviewing, I got a bit confused wondering where `bzero` was defined (because it was all lowercase). So, the rename might make it clearer.

Reply:
Thanks for great suggestion! Fixed in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3390/commits/64508ff17fc3fda2c2c3efc3a2d2bee938c02917